### PR TITLE
feat: create mobile curriculum data on build

### PR DESCRIPTION
**Pull request by Sembauke**
As discussed with Oliver it would be useful to statically generate curriculum data for the mobile client on build.

If we want to make learn natively available in the mobile client, this is the first step.

### CODE REVIEW

File: tools/scripts/build/build-mobile-curriculum.js [hunk 1]
```diff
@@ -0,0 +1,54 @@
+const fs = require('fs');
+const path = require('path');
+
+exports.buildMobileCurriculum = function buildMobileCurriculum(json) {
+  const mobileStaticPath = path.resolve(__dirname, '../../../client/static');
+
+  if (!fs.existsSync(`${mobileStaticPath}/mobile`)) {
+    fs.mkdir(`${mobileStaticPath}/mobile`, err => {
+      return err;
+    });
```
Comment (ojeytonwilliams) [inline tools/scripts/build/build-mobile-curriculum.js:10]:
```suggestion
    fs.mkdirSync(`${mobileStaticPath}/mobile`)
```
It needs to be `mkdirSync` or there will be a race condition between `mkdir` and `writeJson`.  A race that `mkdir` will almost certainly win, but still!

Also, throwing an error is fine here: if this doesn't work I doubt we can recover.

File: tools/scripts/build/build-mobile-curriculum.js [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+    writeJson(json);
+  } else {
+    writeJson(json);
+  }
+
+  function writeJson(curriculum) {
```
Comment (ojeytonwilliams) [inline tools/scripts/build/build-mobile-curriculum.js:16]: writeJson is a bit generic.  Could you name it something specific?

File: tools/scripts/build/build-curriculum.js [hunk 3]
```diff
@@ -2,11 +2,15 @@ const fs = require('fs');
...
 const globalConfigPath = path.resolve(__dirname, '../../../config');
 
 // We are defaulting to English because the ids for the challenges are same
 // across all languages.
 getChallengesForLang('english')
   .then(JSON.stringify)
-  .then(x => fs.writeFileSync(`${globalConfigPath}/curriculum.json`, x));
+  .then(json => {
+    fs.writeFileSync(`${globalConfigPath}/curriculum.json`, json);
+    buildMobileCurriculum(json);
```
Comment (ojeytonwilliams) [inline tools/scripts/build/build-curriculum.js:15]: Can't you avoid parsing the curriculum string by calling `buildMobileCurriculum` after `getChallengesForLang`?

File: tools/scripts/build/build-mobile-curriculum.js [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+    curriculum = JSON.parse(curriculum);
+
+    const superBlocks = Object.keys(curriculum);
+
+    const superBlockObj = { superblocks: superBlocks.slice(1) };
```
Comment (ojeytonwilliams) [inline tools/scripts/build/build-mobile-curriculum.js:21]:
```suggestion
    const superBlocksAndCertifications = Object.keys(curriculum);

    const superBlockObj = { superblocks: superBlocksAndCertifications.slice(1) };
```
I think this is enough to explain why the slice is necessary.

File: tools/scripts/build/build-mobile-curriculum.js [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+      }
+
+      writeToFile(superBlocks[i], blockNames);
+      blockNames = {};
+    }
+  }
+
+  function writeToFile(superblock, json) {
+    fs.writeFileSync(
+      `${mobileStaticPath}/mobile/${superblock}.json`,
```
Comment (ojeytonwilliams) [inline tools/scripts/build/build-mobile-curriculum.js:50]:
```suggestion
  function writeToFile(filename, json) {
    fs.writeFileSync(
      `${mobileStaticPath}/mobile/${filename}.json`,
```

File: tools/scripts/build/build-mobile-curriculum.js [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+      JSON.stringify(json, null, 2)
```
Comment (ojeytonwilliams) [inline tools/scripts/build/build-mobile-curriculum.js:51]:
```suggestion
      JSON.stringify(json)
```
Is there any need to prettify this?

File: tools/scripts/build/build-mobile-curriculum.js [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+
+    const superBlockObj = { superblocks: superBlocks.slice(1) };
+
+    writeToFile('availableSuperblocks', superBlockObj);
+
+    let blockNames = {};
+
+    for (let i = 0; i < superBlocks.length; i++) {
+      // blockGroup = object parent without key-name
+
```
Comment (ojeytonwilliams) [inline tools/scripts/build/build-mobile-curriculum.js:29]:
```suggestion
    for (let i = 0; i < superBlocks.length; i++) {
      // blockGroup = object parent without key-name

      const blockNames = {};
```
Bit neater, imo.

File: tools/scripts/build/build-mobile-curriculum.js [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+      const blockGroup = Object.keys(curriculum[superBlocks[i]].blocks);
```
Comment (ojeytonwilliams) [inline tools/scripts/build/build-mobile-curriculum.js:30]: Shouldn't this be just `blockNames`?  It's an array of them, after all.  I think it's just a straight swap with the current `blockNames`!

**Review by ojeytonwilliams — CHANGES_REQUESTED** [review]
Almost entirely nitpicks:

File: tools/scripts/build/build-curriculum.js (hunk 3 above)
Comment (Sembauke) [reply tools/scripts/build/build-curriculum.js:15]: Not exactly sure what you mean by this.

File: tools/scripts/build/build-mobile-curriculum.js (hunk 7 above)
Comment (Sembauke) [reply tools/scripts/build/build-mobile-curriculum.js:27]: is the blockNames supposed to be a constant?

File: tools/scripts/build/build-mobile-curriculum.js (hunk 8 above)
Comment (Sembauke) [reply tools/scripts/build/build-mobile-curriculum.js:28]: Not sure what you mean by this.

File: tools/scripts/build/build-mobile-curriculum.js (hunk 6 above)
Comment (Sembauke) [reply tools/scripts/build/build-mobile-curriculum.js:49]: could be useful for debugging..

File: tools/scripts/build/build-mobile-curriculum.js (hunk 8 above)
Comment (ojeytonwilliams) [reply tools/scripts/build/build-mobile-curriculum.js:28]: Sorry: I meant that `blockGroup` has this structure

```
[
  'how-neural-networks-work',
  'machine-learning-with-python-projects',
  'tensorflow'
]
```
i.e. just an array of names.  So, it makes sense to call it `blockNames`.  That name is currently being used, though, so you'd need to rename the existing one.

**Review by moT01 — DISMISSED** [review]
LGTM 🎉 Looked over it pretty good 👍

File: tools/scripts/build/build-curriculum.js (hunk 3 above)
Comment (ojeytonwilliams) [reply tools/scripts/build/build-curriculum.js:15]: The current approach takes an object, stringifies it, writes it to a file and then has to parse that string again.  If we call `buildMobileCurriculum` before `JSON.stringify` we don't have to parse the string.

Comment (Sembauke) [reply tools/scripts/build/build-curriculum.js]: This makes sense

**Review by moT01 — APPROVED** [review]

**Review by ojeytonwilliams — APPROVED** [review]

**Pull request merged**